### PR TITLE
GroupBy backpressure fix

### PR DESCRIPTION
**Pull request by akarnokd**
This is a backport of the 2.x GroupBy operator which solves #3425.

One unit test in OperatorRetryTest had to be altered a bit. I believe
the original code relied on a GroupBy behavior which caused the bug in
#3425.

### CODE REVIEW

**Comment by stealthcode** [comment]
The following test starts breaking with this branch. It prints out 1 to 128 then stops and does not print anything else (apparently no progress is made). In branch 1.x this continues to print past 128. 

```java
range(1, RxRingBuffer.SIZE *4).repeat()
    .groupBy((d) -> d)
    .flatMap((go) ->
        go.doOnNext(System.out::println)
            .subscribeOn(Schedulers.io())
    )
    .subscribe();
```

**Comment by akarnokd** [comment]
This is because the lack of horizontal unboundedness of flatMap and the bug in groupBy together made it work in 1.0.14. If I run this code with the 2.x fix and add 1024 as the max concurrency value, it prints lots of values indefinitely.

File: src/test/java/rx/internal/operators/OperatorGroupByTest.java [hunk 1]
```diff
@@ -1501,4 +1480,103 @@ public void onNext(Integer t) {
...
+                    public Object call(Integer v) {
+                        return v;
+                    }
+                })
+                .doOnNext(new Action1<GroupedObservable<Object, Integer>>() {
+                    @Override
+                    public void call(GroupedObservable<Object, Integer> g) {
+                        g.subscribe();
+                    }
+                }) // this will request Long.MAX_VALUE
```
Comment (stealthcode) [inline src/test/java/rx/internal/operators/OperatorGroupByTest.java]: Is this comment meant for the inner observable? The outer observable gets a request of 128.

**Comment by akarnokd** [comment]
I've updated the comments.

**Comment by stealthcode** [comment]
:+1:

**Pull request merged**